src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const statesHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    statesHistory.push(currentState);
```

Comment:
It is better to push a copy of the object to the array at the end of each loop cycle to prevent any accidental mutations from affecting the states in `statesHistory`. However, you are already doing this correctly by pushing `currentState` after the switch block, which is a new object at this point.